src/index.ts

```diff
@@ -50,73 +51,75 @@ const convertNodeUrl = ({
-    newQuery: Update<ParsedUrl['query']>,
-): MapParsedUrlFn => parsedUrl => ({
+export const replaceQueryInParsedUrl = (newQuery: Update<ParsedUrl['query']>): MapParsedUrlFn => (
+    parsedUrl,
+) => ({
     ...parsedUrl,
-    query: newQuery instanceof Function ? newQuery(parsedUrl.query) : newQuery,
+    query: typeof newQuery === 'function' ? newQuery(parsedUrl.query) : newQuery,
 });
 
-export const replaceQueryInUrl = pipe(
-    replaceQueryInParsedUrl,
-    mapUrl,
-);
+export const replaceQueryInUrl = pipe(replaceQueryInParsedUrl, mapUrl);
 
 export const addQueryToParsedUrl = (queryToAppend: ParsedUrl['query']): MapParsedUrlFn =>
-    replaceQueryInParsedUrl(existingQuery => ({ ...existingQuery, ...queryToAppend }));
+    replaceQueryInParsedUrl((prevQuery) => ({ ...prevQuery, ...queryToAppend }));
 
-export const addQueryToUrl = pipe(
-    addQueryToParsedUrl,
-    mapUrl,
-);
+export const addQueryToUrl = pipe(addQueryToParsedUrl, mapUrl);
 
-type ParsedPath = Pick<ParsedUrl, 'query' | 'pathname'>;
+interface ParsedPath extends Pick<ParsedUrl, 'query' | 'pathname'> {}
 
 const parsePath = pipe(
     parseUrlWithQueryString,
     ({ query, pathname }): ParsedPath => ({ query, pathname }),
 );
 
-type Path = Required<NodeUrlObjectWithParsedQuery>['path'];
+const parseNullablePath = pipe(
+    mapMaybe(parsePath),
+    getOrElseMaybe((): ParsedPath => ({ query: null, pathname: null })),
+);
 
-const parseNullablePath = (maybePath: Path): ParsedPath =>
-    pipeWith(
-        maybePath,
-        maybe => mapMaybe(maybe, parsePath),
-        maybe => getOrElseMaybe(maybe, () => ({ query: {}, pathname: null })),
-    );
+const formatPath = (parsedPath: ParsedPath) => urlHelpers.format(parsedPath);
+
+type Path = urlHelpers.Url['path'];
+
+const pathCodec: Codec<ParsedPath, Path> = {
+    decode: parseNullablePath,
+    encode: formatPath,
+};
 
 const convertUpdatePathFnToUpdateParsedPathFn = (
     updatePath: UpdateFn<Path>,
-): UpdateFn<ParsedPath> =>
-    pipe(
-        urlHelpers.format,
-        updatePath,
-        parseNullablePath,
-    );
+): UpdateFn<ParsedPath> => pipe(pathCodec.encode, updatePath, pathCodec.decode);
 
-const convertUpdatePathToUpdateParsedPath = (newPath: Update<Path>) =>
-    newPath instanceof Function
+const convertUpdatePathToUpdateParsedPath = (newPath: Update<Path>): Update<ParsedPath> =>
+    typeof newPath === 'function'
         ? convertUpdatePathFnToUpdateParsedPathFn(newPath)
         : parseNullablePath(newPath);
 
-export const replacePathInParsedUrl = (newPath: Update<ParsedPath>): MapParsedUrlFn => parsedUrl =>
-    pipeWith(newPath instanceof Function ? newPath(parsedUrl) : newPath, newPathParsed => ({
-        ...parsedUrl,
-        ...newPathParsed,
-    }));
+export const replacePathInParsedUrl = (newPath: Update<ParsedPath>): MapParsedUrlFn => (
+    parsedUrl,
+) => ({
+    ...parsedUrl,
+    ...(typeof newPath === 'function' ? newPath(parsedUrl) : newPath),
```

Comment:
We're doing these checks in a bunch of places. Maybe worth extracting that out? In case we need to change it again.

Reply:
That's essentially already done in the `mococle-ts` PR: https://github.com/unsplash/url-transformers/pull/6